File: customer_service/frontend/customer-service-chatgpt-faq-app/src/app/models/streaming.models.ts
export interface StreamEvent {
  type: string;
  data: any;
  timestamp: string;
}

export interface PredictionEvent extends StreamEvent {
  type: 'prediction';
  data: {
    predicted_tools: string[];
    confidence: number;
    prediction_method: string;
  };
}

export interface AuthCheckEvent extends StreamEvent {
  type: 'auth_check';
  data: {
    required: boolean;
    valid: boolean;
    protected_tools: string[];
    user_id?: string;
  };
}

export interface ToolExecutionStartEvent extends StreamEvent {
  type: 'tool_execution_start';
  data: {
    tool_name: string;
    arguments: any;
    step: number;
    total_steps: number;
  };
}

export interface ToolExecutionCompleteEvent extends StreamEvent {
  type: 'tool_execution_complete';
  data: {
    tool_name: string;
    result: string;
    duration_ms: number;
    success: boolean;
    step: number;
  };
}

export interface ProgressEvent extends StreamEvent {
  type: 'progress';
  data: {
    total_steps: number;
    completed_steps: number;
    current_step: string;
    percentage: number;
  };
}

export interface FinalAnswerEvent extends StreamEvent {
  type: 'final_answer';
  data: {
    response: string;
    summary: StreamingSummary;
    execution_time_ms: number;
    tools_used: string[];
    prediction_accuracy: boolean;
    error_count: number;
  };
}

export interface ErrorEvent extends StreamEvent {
  type: 'error';
  data: {
    error: string;
    step: string;
    critical?: boolean;
    tool_name?: string;
  };
}

export interface StreamingSummary {
  query: string;
  predicted_tools: string[];
  actual_tools_used: string[];
  prediction_accuracy: boolean;
  total_execution_time_ms: number;
  tools_executed: ToolExecutionInfo[];
  conversation_steps: any[];
  final_response: string;
  error_count: number;
  warnings?: string[];
}

export interface ToolExecutionInfo {
  tool_name: string;
  arguments: any;
  start_time: string;
  end_time?: string;
  result?: any;
  error?: string;
  duration_ms?: number;
}

export interface ToolProgress {
  step: number;
  total: number;
  current: string;
  percentage: number;
}

export interface StreamingResponse {
  type: 'event' | 'complete' | 'error';
  event?: StreamEvent;
  error?: string;
  timestamp: Date;
}

export interface StreamingQueryRequest {
  query: string;
  history?: any[];
  include_intermediate?: boolean;
  include_progress?: boolean;
}

export interface ChatMessage {
  id?: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: Date;
  streaming?: boolean;
  streamEvents?: StreamEvent[];
  toolsUsed?: string[];
  executionTime?: number;
  error?: boolean;
}

export interface StreamingChatState {
  isStreaming: boolean;
  currentProgress: ToolProgress;
  currentEvents: StreamEvent[];
  error?: string;
}

File: customer_service/backend/streaming_models.py
from datetime import datetime
from typing import Any, Dict, List, Literal, Optional, Union
from pydantic import BaseModel, Field


class StreamEvent(BaseModel):
    """Base class for all streaming events"""
    type: str
    data: Dict[str, Any]
    timestamp: datetime = Field(default_factory=datetime.now)


class PredictionEvent(StreamEvent):
    """Event sent when tool prediction is complete"""
    type: Literal["prediction"] = "prediction"


class AuthCheckEvent(StreamEvent):
    """Event sent when authentication check is complete"""
    type: Literal["auth_check"] = "auth_check"


class ToolExecutionStartEvent(StreamEvent):
    """Event sent when a tool starts executing"""
    type: Literal["tool_execution_start"] = "tool_execution_start"


class ToolExecutionCompleteEvent(StreamEvent):
    """Event sent when a tool completes execution"""
    type: Literal["tool_execution_complete"] = "tool_execution_complete"


class ConversationStepEvent(StreamEvent):
    """Event sent for each step in the agent conversation"""
    type: Literal["conversation_step"] = "conversation_step"


class FinalAnswerEvent(StreamEvent):
    """Event sent with the final aggregated response"""
    type: Literal["final_answer"] = "final_answer"


class ErrorEvent(StreamEvent):
    """Event sent when an error occurs during streaming"""
    type: Literal["error"] = "error"


class ProgressEvent(StreamEvent):
    """Event sent to update overall progress"""
    type: Literal["progress"] = "progress"


# Union type for all possible stream events
StreamEventType = Union[
    PredictionEvent,
    AuthCheckEvent,
    ToolExecutionStartEvent,
    ToolExecutionCompleteEvent,
    ConversationStepEvent,
    FinalAnswerEvent,
    ErrorEvent,
    ProgressEvent
]


class StreamingQueryRequest(BaseModel):
    """Request model for streaming queries"""
    query: str
    history: List[Dict[str, str]] = Field(default_factory=list)
    include_intermediate: bool = True
    include_progress: bool = True


class ToolExecutionInfo(BaseModel):
    """Information about tool execution"""
    tool_name: str
    arguments: Dict[str, Any]
    start_time: datetime
    end_time: Optional[datetime] = None
    result: Optional[Any] = None
    error: Optional[str] = None
    duration_ms: Optional[int] = None


class StreamingProgress(BaseModel):
    """Progress information for streaming execution"""
    total_steps: int
    completed_steps: int
    current_step: str
    percentage: int = Field(ge=0, le=100)


class ConversationStep(BaseModel):
    """Individual step in the agent conversation"""
    step_number: int
    agent_name: str
    message_type: str  # "function_call", "function_result", "assistant_response"
    content: str
    function_name: Optional[str] = None
    function_args: Optional[Dict[str, Any]] = None
    function_result: Optional[Any] = None


class StreamingSummary(BaseModel):
    """Summary of the streaming execution"""
    query: str
    predicted_tools: List[str]
    actual_tools_used: List[str]
    prediction_accuracy: bool
    total_execution_time_ms: int
    tools_executed: List[ToolExecutionInfo]
    conversation_steps: List[ConversationStep]
    final_response: str
    error_count: int = 0
    warnings: List[str] = Field(default_factory=list)